lib/AST/ASTContext.cpp

```diff
@@ -2386,23 +2393,56 @@ getModuleVersionKindString(const ModuleLoader::ModuleVersionInfo &info) {
   }
 }
 
-bool ASTContext::canImportModuleImpl(ImportPath::Module ModuleName,
-                                     llvm::VersionTuple version,
-                                     bool underlyingVersion,
-                                     bool updateFailingList) const {
+void ASTContext::addSucceededCanImportModule(
+    StringRef moduleName, bool underlyingVersion,
+    const llvm::VersionTuple &versionInfo) {
+  auto &entry = SucceededModuleImportNames[moduleName.str()];
+  if (!versionInfo.empty()) {
+    if (underlyingVersion)
+      entry.UnderlyingVersion = versionInfo;
```

Comment:
What if `canImport()` is evaluated multiple times with different version numbers? Shouldn't we only store the new version if it is higher than the previously stored version?

Reply:
The version stored is the version of the module, not the version in `canImport` statement. The version is either requested or not requested, and cannot change during the build. And we can use that one version to check for every version requested through `canImport`.

Reply:
I see. I find this naming ("Succeeded") confusing then; these are just the versions for these modules. It's just a nit, though.

Reply:
I will rename it.